nuxeo-features/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java

```diff
@@ -0,0 +1,352 @@
+            if (error != null) {
+                throw new NuxeoException(error, Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
+            }
+            String resURL = StringUtils.stripEnd(ctx.getURL(), "/status");
+            return redirect(resURL);
+        } else {
+            Object result = "RUNNING";
+            if (isAsync(executionId)) {
+                Serializable commandId = getCommandId(executionId);
+                result = getAsyncService(executionId).getStatus(commandId);
+            }
+            return ResponseHelper.getResponse(result, request, Response.Status.OK.getStatusCode());
+        }
+    }
+
+    @GET
+    @Path("{executionId}")
+    public Object result(@PathParam("executionId") String executionId)
+            throws IOException, URISyntaxException, MessagingException {
+
+        if (isCompleted(executionId)) {
+            Object output = getResult(executionId);
+
+            String error = getError(executionId);
+
+            // cleanup after result is accessed
+            cleanup(executionId);
+
+            if (error != null) {
+                throw new NuxeoException(error, Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
+            }
+
+            // if output has a "url" key assume it's a redirect url
+            if (output instanceof Map) {
+                Object url = ((Map) output).get("url");
+                if (url instanceof String) {
+                    String baseUrl = VirtualHostHelper.getBaseURL(ctx.getRequest());
+                    return redirect(baseUrl + url);
+                }
+            }
+            return ResponseHelper.getResponse(output, request, Response.Status.OK.getStatusCode());
+        }
+
+        throw new WebResourceNotFoundException("Execution with id=" + executionId + " not found");
+    }
+
+    @DELETE
+    @Path("{executionId}")
+    public Object abort(@PathParam("executionId") String executionId) throws IOException, MessagingException {
+        if (exists(executionId) && !isCompleted(executionId)) {
+            // XXX: only supports abort of async commands
+            if (isAsync(executionId)) {
+                Serializable commandId = getCommandId(executionId);
+                return getAsyncService(executionId).abort(commandId);
+            }
+            return ResponseHelper.getResponse("RUNNING", request, Response.Status.OK.getStatusCode());
+        }
+        throw new WebResourceNotFoundException("Execution with id=" + executionId + " has completed");
+    }
+
+    private TransientStore getTransientStore() {
+        return Framework.getService(TransientStoreService.class).getStore(STATUS_STORE_NAME);
+    }
+
+    private void enterMethod(String executionId, InvokableMethod method) {
+        Map<String, Serializable> parameters = new HashMap<>();
+        // AsyncService.class is default => not async
+        if (!AsyncService.class.equals(method.getAsyncService())) {
```

Comment:
Should'n be `isAssignableFrom` ?

Reply:
Perhaps the comment is more confusing than helpful. I have to use a default value for the Class in the interface of the OperationMethod annotation so I went with AsyncService.class.
`Class<? extends AsyncService> asyncService() default AsyncService.class;`
This means that here I need to check if it really isn't AsyncService.class. Given the annotation signature it must be assignable to AsyncService for it to compile.